### PR TITLE
add: support to parse and store features with multiple locations; modify final feature sort

### PR DESCRIPTION
**Pull request by mohitdmak**
## Changes:
* Changes in the GenBank -> SBOL3 converter to support storing multiple FeatureLocations for each Feature are very straightforward:
    - Instead of parsing only the 1st FeatureLocation of Features from GenBank, I have created an array of FeatureLocations, to store all parsed FeatureLocations. 
    - Also, for this direction of conversion, I have not ensured any kind of sorting for Features or FeatureLocations, and am not sure if it is even required.
* For the SBOL3 -> GenBank direction of conversion:
    1. For each Feature, `feat_loc` shall hold the final FeatureLocation object *(which will be just `FeatureLocation` if there is only 1 location, else `CompoundLocation`, if there are multiple)*.
    2. More info on `CompoundLocation` can be found [here](https://biopython.org/docs/1.75/api/Bio.SeqFeature.html#Bio.SeqFeature.CompoundLocation).
    3. For each Feature, `feat_loc_array` will be the "flattened" list of start and end points of all locations of a feature, sorted internally:
        - Eg: For a feature having locations - [2,4], [3,9], [1,30]; the `feat_loc_array` will hold [1,30,2,4,3,9]
        - This will be the basis for performing sorting **among** different features.
    4. Also, `feat_order` is a dictionary that will have key-value pairs of *Feature Object - `feat_locs` array*. While performing the final sorting of features, this dictionary is used to access the `feat_locs` array for each feature.
    5. In case a Feature has multiple locations, it will be stored in a `feat_locs` array, which will be internally sorted for each Feature on the basis of each location's start and end points lexicographically before being stored in the final Feature object.
* The unit test functions have been refactored, but not as well as [this PR](https://github.com/SynBioDex/SBOL-utilities/pull/158) has. I can either push the refactoring commits from that PR here or wait for it to be merged.  

## TODO: 
1. The `CompoundLocation` Object takes in an **operator** argument which can be either of "join"/"order"/"compliment"; currently, I am using "join" whenever there are multiple locations. We need to discuss how the converter should decide which one to use based on parsed feature locations' details.
2. The multiple location converter works, for the most part, except for when parsing the feature with label **"c5"** in the outer `sequence2_modified.gb` file. This feature has location of the form:
```
complement(join(1603..1891,2322..2438,2538..2633,2801..2843,
                     2918..3073,3167..3247,3874..3972,4082..4637))
```
But once this is round-tripped to return the `sequence2_modified.gb` file inside the `test/test_files` directory, the location becomes:
```
complement(join(4082..4637,3874..3972,3167..3247,
                     2918..3073,2801..2843,2538..2633,2322..2438,1603..1891))
```
The order inside the join operator reverses, and I am struggling to understand how to deal with this issue. I think this is related to the previous point, and maybe the operator used should not be "join" here.

### CODE REVIEW

File: sequence2_modified.gb [hunk 1]
```diff
@@ -0,0 +1,161 @@
+LOCUS       AF165912                5485 bp    DNA     linear   PLN 29-JUL-1999
```
Comment (jakebeal) [inline sequence2_modified.gb:1]: This looks like an accidental check-in duplicate of the test file.

**Review by jakebeal — CHANGES_REQUESTED** [review]

**Comment by jakebeal** [comment]
Regarding the reverse complement question: do the features need to be sorted in the opposite order when it's on the opposite strand?

**Review by tcmitchell — CHANGES_REQUESTED** [review]
Requested changes:
* Remove top level file `sequence2_modified.gb`
* Remove debug prints
* Work with @jakebeal on the ordering of multiple locations

Comment, not a requested change:
* See the comment about combining two similar paths when handling feature locations

File: sbol_utilities/sbol3_genbank_conversion.py [hunk 2]
```diff
@@ -269,20 +272,49 @@ def convert_sbol3_to_genbank(self, sbol3_file: str, doc: sbol3.Document = None,
...
+                                    # TODO: Raise custom converter class ERROR for `else:`
+                                    feat_loc = FeatureLocation(
+                                        start=obj_feat_loc.start,
+                                        end=obj_feat_loc.end,
+                                        strand=feat_strand,
+                                    )
+                                    feat_locs.append(feat_loc)
+                                # sort feature locations lexicographically internally first
+                                feat_locs.sort(key=lambda loc: (loc.start, loc.end))
+                                print(f"locs {feat_locs}")
```
[the remark this answers is not in the document]
Comment (mohitdmak) [reply sbol_utilities/sbol3_genbank_conversion.py]: Done

File: sbol_utilities/sbol3_genbank_conversion.py [hunk 3]
```diff
@@ -269,20 +272,49 @@ def convert_sbol3_to_genbank(self, sbol3_file: str, doc: sbol3.Document = None,
...
-                                feat_strand = self.BIO_STRAND_REVERSE
-                            elif obj_feat_loc.orientation != sbol3.SO_FORWARD:
-                                raise ValueError(f"Location orientation: `{obj_feat_loc.orientation}` for feature: `{obj_feat.name}` of component: `{obj.display_id}` is not a valid orientation.\n Valid orientations are `{sbol3.SO_FORWARD}`, `{sbol3.SO_REVERSE}`")
-                            # TODO: Raise custom converter class ERROR for `else:`
-                            feat_loc = FeatureLocation(
-                                start=obj_feat_loc.start,
-                                end=obj_feat_loc.end,
-                                strand=feat_strand,
-                            )
+                            if len(obj_feat.locations) == 1:
```
[the remark this answers is not in the document]
Comment (mohitdmak) [reply sbol_utilities/sbol3_genbank_conversion.py]: Thanks for suggesting this; the combined code looks much better, I have pushed this in the latest commit. I have renamed a few variables in a more intuitive way.

File: sequence2_modified.gb (hunk 1 above)
Comment (mohitdmak) [reply sequence2_modified.gb]: Yep, removed.

**Comment by mohitdmak** [comment]
> Regarding the reverse complement question: do the features need to be sorted in the opposite order when it's on the opposite strand?

Thanks, this works, I have added a check to see if the orientation on an SBOL3 feature is `sbol3.SO_REVERSE`, in which case, the sort needs to be in the reverse order.

**Review by tcmitchell — APPROVED** [review]
Thanks for incorporating the feedback!

**Review by jakebeal — APPROVED** [review]

**Pull request merged**